source/common/http/conn_manager_impl.cc

```diff
@@ -1767,6 +1764,12 @@ void ConnectionManagerImpl::ActiveStream::recreateStream(
   connection_manager_.doEndStream(*this, /*check_for_deferred_close*/ false);
 
   RequestDecoder& new_stream = connection_manager_.newStream(*response_encoder, true);
+  // Set the new RequestDecoder on the ResponseEncoder. Even though all of the decoder callbacks
+  // have already been called at this point, the encoder still needs the new decoder for deferred
+  // logging in some cases.
+  // This doesn't currently work for HTTP/1 as the H/1 ResponseEncoder doesn't hold the active
+  // stream's pointer to the RequestDecoder.
```

Comment:
For my own knowledge, how does H1 codec decode request headers in that case?

Reply:
H1 has an [ActiveStream](https://github.com/envoyproxy/envoy/blob/28ac8d08d71b41544969d23415247dc53ce919c1/source/common/http/http1/codec_impl.h#L465-L466) that has a handle on both a ResponseEncoder and RequestDecoder. So e.g. on receiving headers the connection goes through the active request to get to the decoder, like [here](https://github.com/envoyproxy/envoy/blob/28ac8d08d71b41544969d23415247dc53ce919c1/source/common/http/http1/codec_impl.cc). It seems like the H/2 and H/3 codecs have the decoding path go through the response encoder -> request decoder partly to simplify multiplexing, so it makes sense that H/1 (which doesn't have streams) is organized a bit differently. I could be wrong about that last part.

Reply:
Get it! Thanks for the link!